homeassistant/components/zwave_mqtt/__init__.py

```diff
@@ -0,0 +1,332 @@
+
+            values = ZWaveDeviceEntityValues(hass, options, schema, value)
+            values.setup()
+
+            # We create a new list and update the reference here so that
+            # the list can be safely iterated over in the main thread
+            data_values[node_id] = node_data_values + [values]
+
+    @callback
+    def async_value_changed(value):
+        # if an entity belonging to this value needs updating,
+        # it's handled within the entity logic
+        _LOGGER.debug(
+            "[VALUE CHANGED] node_id: %s - label: %s - value: %s - value_id: %s - CC: %s",
+            value.node.id,
+            value.label,
+            value.value,
+            value.value_id_key,
+            value.command_class,
+        )
+        # Handle a scene activation message
+        if value.command_class in [
+            CommandClass.SCENE_ACTIVATION,
+            CommandClass.CENTRAL_SCENE,
+        ]:
+            handle_scene_activated(hass, value)
+            return
+
+    @callback
+    def async_value_removed(value):
+        _LOGGER.debug(
+            "[VALUE REMOVED] node_id: %s - label: %s - value: %s - value_id: %s - CC: %s",
+            value.node.id,
+            value.label,
+            value.value,
+            value.value_id_key,
+            value.command_class,
+        )
+        # signal all entities using this value for removal
+        value_unique_id = f"{value.node.id}-{value.value_id_key}"
+        async_dispatcher_send(hass, const.SIGNAL_DELETE_ENTITY, value_unique_id)
+        # remove value from our local list
+        node_data_values = data_values[value.node.id]
+        node_data_values[:] = [
+            item for item in node_data_values if item.values_id != value_unique_id
+        ]
+
+    # Listen to events for node and value changes
+    options.listen(EVENT_NODE_ADDED, async_node_added)
+    options.listen(EVENT_NODE_CHANGED, async_node_changed)
+    options.listen(EVENT_NODE_REMOVED, async_node_removed)
+    options.listen(EVENT_VALUE_ADDED, async_value_added)
+    options.listen(EVENT_VALUE_CHANGED, async_value_changed)
+    options.listen(EVENT_VALUE_REMOVED, async_value_removed)
+    options.listen(EVENT_INSTANCE_EVENT, async_instance_event)
+
+    # Register Services
+    services = ZWaveServices(hass, manager, data_nodes)
+    services.register()
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    # cleanup platforms
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if not unload_ok:
+        return False
+
+    # unsubscribe all listeners
+    for unsubscribe_listener in hass.data[DOMAIN][entry.entry_id][DATA_UNSUBSCRIBE]:
+        unsubscribe_listener()
+    hass.data[DOMAIN][entry.entry_id][DATA_UNSUBSCRIBE].clear()
```

Comment:
No need to clear if we're popping the whole entry from the dict.

Reply:
Commit done.